src/scoped.rs

```diff
@@ -233,26 +252,16 @@ impl<'a> Scope<'a> {
     // method outside of any destructor, we avoid any leakage problems
     // due to @rust-lang/rust#14875.
     fn drop_all(&mut self) {
-        loop {
-            // use a separate scope to ensure that the RefCell borrow
-            // is relinquished before running `dtor`
-            let dtor = {
-                let mut dtors = self.dtors.borrow_mut();
-                if let Some(mut node) = dtors.take() {
-                    *dtors = node.next.take().map(|b| *b);
-                    node.dtor
-                } else {
-                    return;
-                }
-            };
-            dtor.call_box()
+        while let Some(dtor) = DtorChain::pop(&mut self.dtors.borrow_mut()) {
+            dtor.call_box();
```

Comment:
Consider this:

```rust
panic::catch_unwind(|| {
    let foo = "foo".to_string();
    crossbeam::scope(|s| {
        s.spawn(|| panic!(foo));
        s.spawn(|| panic!(foo));
    });
});
```

Let's say we call `dtor.call_box()` here to wait for the second thread and it panics due to `.join().unwrap()` inside the boxed closure. We catch the panic with `catch_unwind`. But look - we haven't joined on the first thread yet, which might be still holding a reference to the now destroyed `foo`. This is undefined behavior.

Unless I'm mistaken, the right thing would be to catch any potential panics coming from `dtor.call_box()` and make sure to join on all threads. Only after all threads have finished, if there was a panic, re-throw the first panic we found and keep unwinding.

Reply:
@stjepang stupid question: I was wondering how to exit the main thread if at least one thread panics?
i.e. in the above example panic just after printing 2.
Thanks!

Reply:
@k3yavi The main thread *must* wait for all child threads to finish, or else its stack-allocated data might get destroyed before the child threads stop using it.

We definitely need some way of propagating panics from child threads onto the main thread, though.